server/openslides/poll/access_permissions.py

```diff
@@ -71,8 +72,21 @@ async def get_restricted_data(
         """
 
         # add has_voted for all users to check whether op has voted
+        # also fill user_has_voted_for_delegations with all users for which he has
+        # already voted
+        user_data = await element_cache.get_element_data(
```

Comment:
This does currently not work (sometimes, with the right (wrong?) conditions):

First, do not raise errors here. The restricters should not have a reason to have invalid data and there is no other reason to throw errors. So you have to expect the user_data to be available. Proposal: Do an error log here and set `vote_delegated_from_ids` to `[]` and proceed.

The seconds, more subtile but sevre part is the cache building. If the cache is empty, it will be filled (cache.py:125). Since last year all models were independent, so the order is not important. The code can build motions first and then the categories or vice versa. With the vote-weight this changed: https://github.com/OpenSlides/OpenSlides/blob/master/server/openslides/poll/models.py#L227

This line caused a lot of trouble because now we have a dependency from */*_poll to core/config. This resulted in this commit: https://github.com/OpenSlides/OpenSlides/commit/b78372f8a3eb51c5c4e1c5753f500a46416900a0

I manually build the config first, so it is available (the `config[key]` accesses the `element_cache`). With the marked line, we now have a dependency from */*_poll to user. In theory this is not a problem as we have a tree as the dependency graph, but this is not implemented in OS3 yet, since we just didn't had dependencies on serialization, except the one for config. But the config was always a bit special...

Note: Normally We surrounded this problem by using serializermethodfields, but here I see no possiblity to inject `vote_delegated_from_users_id` from all users into all polls only to remove the dependency.

-> I think that your approach is good and works, but we have to manage the dependencies. Proposal: Hard-code the dependencies in the cache.py. The nicer approach would be something like done in the migrations, that we write into a model class that it needs another model class before objects can be serialized. Then the build cache first does the config, than users, than the rest.

Note that this not the only place where models are serialized: A wrong order can potentially cause problems e.g. here: https://github.com/OpenSlides/OpenSlides/blob/master/server/openslides/utils/autoupdate.py#L107 (I wonder myself why I didn't care about this case back then; I just overlooked I guess).

Please think about it and maybe we should speak about possible solutions.

Reply:
Another after-thought: When updating a user (or just the `vote_delegated_from_users_id` field), you have to re-serialize all polls. The same is also true vor the vote weight config, but I guess this was missed, too. In fact that we need this information I wonder, if we can put it into the user model instead of polls. This solves the user-update problem, but then there must be autoupdates for the voted user, when a vote is received (more autoupdates here).

Reply:
Ok, it is not as severe as I though: Since it is in the restricters this code is actually not called when the cache is built. These fields even doesn't exist there. The config-thing is serialization-related, but this access to the cache is restricter-related. But this needs to be watched carefully...